src/pad.rs

```diff
@@ -0,0 +1,363 @@
+
+        let mut db = [0u8; MAX_MLEN];
+
+        let db_len = em_len - digest_len - 1;
+        let pad_len = db_len - self.s_len - 1;
+
+        for i in 0..pad_len {
+            db[i] = 0;
+        }
+        db[pad_len] = 1;
+        db[pad_len + 1..][..self.s_len].copy_from_slice(&salt[..self.s_len]);
+
+        //    9.  Let dbMask = MGF(H, emLen - hLen - 1).
+
+        //    10. Let maskedDB = DB \xor dbMask.
+
+        let mut db_mask = [0u8; MAX_MLEN];
+
+        try!((self.mgf)(h_hash.as_ref(), db_len, &mut db_mask, self.digest_alg));
+
+        for i in 0..db_len {
+            db[i] ^= db_mask[i];
+        }
+
+        //    11. Set the leftmost 8emLen - emBits bits of the leftmost octet in
+        //        maskedDB to zero.
+
+        //    12. Let EM = maskedDB || H || 0xbc.
+
+        //    13. Output EM.
+
+        out[..db_len].copy_from_slice(&db[..db_len]);
+        out[db_len..][..digest_len].copy_from_slice(h_hash.as_ref());
+        out[db_len + digest_len] = 0xbc;
+
+        Ok(())
+    }
+
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input)
+              -> Result<(), ()> {
+
+        let em_len = encoded.len();
+        encoded.read_all((), |em| {
+            let digest_len = self.digest_alg.output_len;
+
+            // Steps:
+
+            //    1.  If the length of M is greater than the input limitation for the
+            //        hash function (2^61 - 1 octets for SHA-1), output "inconsistent"
+            //        and stop.
+
+            //    2.  Let mHash = Hash(M), an octet string of length hLen.
+
+            let m_hash = digest::digest(self.digest_alg, msg.as_slice_less_safe());
+
+            //    3.  If emLen < hLen + sLen + 2, output "inconsistent" and stop.
+
+            //    4.  If the rightmost octet of EM does not have hexadecimal value
+            //        0xbc, output "inconsistent" and stop.
+
+            //    5.  Let maskedDB be the leftmost emLen - hLen - 1 octets of EM, and
+            //        let H be the next hLen octets.
+            let db_len = em_len - digest_len - 1;
+            let masked_db = try!(em.skip_and_get_input(db_len));
+            let h_hash = try!(em.skip_and_get_input(digest_len));
+
+            if try!(em.read_byte()) != 0xbc {
+                return Err(());
+            }
+
+            //    6.  If the leftmost 8emLen - emBits bits of the leftmost octet in
+            //        maskedDB are not all equal to zero, output "inconsistent" and
+            //        stop.
+
+            //    7.  Let dbMask = MGF(H, emLen - hLen - 1).
+            let mut db = [0u8; MAX_MLEN];
+
+            try!((self.mgf)(h_hash.as_slice_less_safe(), db_len, &mut db, self.digest_alg));
+
+            //    8.  Let DB = maskedDB \xor dbMask.
+
+            try!(masked_db.read_all((), |masked_bytes|{
+                // XXX: First byte may be incorrect due to misaligned number of
+                // bits. This is currently not handled properly
```

Comment:
Related to comment below in the `sign` function. PSS padding is dependent on the number of bits in the modulus.

We would need to add an input variable to this function to specify how long in bits the padding is.